### PR TITLE
!refactor: minimize image complexity

### PR DESCRIPTION
**Pull request by JonZeolla**
# Contributor Comments

Create more, smaller, less complex images by containing less tools and shorter configs in the resulting images.  Allow arbitrary reuse of components through composable docker images, and make more testable components by having each tool be able to be built independently with no other moving parts for troubleshooting or testing.  The final result looks like each tool or environment having a docker file as well as a docker frags and then using jinja2 to allow the combination of those to result in a working image using multi-stage doctor containers.

Also simplify the naming of variables throughout the project by removing the idea of variant(s) and stage(s) and replacing it with more specific and accurate terms like tags, images, or tools, security tools, and environments to simplify future development.

- [x] In `functions.j2` inside of the macro there is no difference between `tools` and `security_tools`, so these should be unified.  Done in 0afa1a0
- [x] A valid version is `2022.05.01`, an invalid version is `2022.05.01-aws`, `v2022.05.01-aws`, or `v2022.05.01`.
- [x] A valid tag is `2022.05.01-aws` but an invalid tag is `v2022.05.01-aws` or `:2022.05.01-aws`.
- [x] A valid tool is `terraform`, `ansible`, etc.
- [x] A valid security_tool is `checkov` or `kics`, etc.
- [X] A valid package is `terraform`, `checkov`, or `aws-cli`.
- [X] A valid alias is `aws` for the package `aws-cli`.
- [X] A valid command is `checkov scan` or `terraform apply`.
- [x] A valid environment is `aws` or `azure`.
- [X] Images are minimal by default.
- [x] The only valid image is `seiso/easy_infra`; `easy_infra` is invalid.
- [x] Combinations are allowed; `image_and_tag` is valid as `seiso/easy_infra:2022.05.01-aws`.
- [x] Migrate stages to use another variable outlined above.
- [x] Migrate variants to use another variable outlined above.
- [X] Documentation is aligned with the code.
- [x] Ensure that the JSON report paths are all created, env vars are set, and outputs go in the right place (Move to easy_infra.yml customizations?  Check `functions.j2` logic to ensure it can handle non-json customizations).
- [x] Confirm testing parity.  Prior to refactor, we had:
  - `minimal`: 71 e2e terraform tests, 19 e2e ansible-playbook tests, and security testing
  - `aws`: 2 integration tests, 71 e2e terraform tests, 19 e2e ansible-playbook tests, and security testing
  - `az`: 2 integration tests, 71 e2e terraform tests, 19 e2e ansible-playbook tests, and security testing
  - "all": 12 interactive path tests, 12 non-interactive path tests, 9 integration tests, 19 e2e ansible-playbook tests, 2 (more?) integration tests, and security testing. (no terraform tests? Seems like an oversight)
- [x] Clean up TODOs and move items to backlog as needed.

## Dependencies
- [x] https://github.com/SeisoLLC/easy_infra/pull/185

## Pull Request Checklist

Thank you for submitting a contribution to `easy_infra`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:

- [X] Rebase your branch against the latest commit of the target branch
- [X] If you are adding a dependency, please explain how it was chosen
- [X] If manual testing is needed in order to validate the changes, provide a testing plan and the expected results
- [X] If there is an issue associated with your Pull Request, link the issue to the PR.

### CODE REVIEW

**Comment by JonZeolla** [comment]
Rebased on main

**Comment by JonZeolla** [comment]
This was just rebased on 4af99ac33665951b8e1307050dd79f5d8871632f and all of the changes were ported

**Comment by JonZeolla** [comment]
Rebased and portd changes up to ef95418. All up to date - now time to get this wrapped up :D

File: tasks.py [hunk 1]
```diff
@@ -3,51 +3,59 @@
...
 from bumpversion.cli import main as bumpversion
 from invoke import task
 
-from easy_infra import __project_name__, __version__, constants, utils
+from easy_infra import __project_name__, __version__, config, constants, utils
 from tests import test as run_test
 
-if platform.machine() == "arm64":
-    PLATFORM: Union[str, None] = "linux/arm64/v8"
+# TODO: Does this work on intel?
```
Comment (JonZeolla) [inline tasks.py:26]: I need someone with an intel-based system to run this @digeseiso / @seanctech

**Comment by JonZeolla** [comment]
Manually confirmed that we have testing parity before/after the refactor.  Testing numbers overall went down due to (1) a reduction in duplication, and (2) a removal of KICS from the terraform image, so I was able to remove all of the KICS-specific terraform tests.  Overall this has more tests, and more specific tests, and those tests should be dynamic based on `easy_infra.yml`.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -45,3 +45,29 @@ pipenv run invoke build --trace
 ```
 
 This will add additional troubleshooting tools to the container, and perform some tracing, putting the details in `/tmp/`.
+
+## High-Level Design
```
Comment (JonZeolla) [inline CONTRIBUTING.md:49]: This should document the high level design that this refactor puts in place.

File: build/hooks/40-set_default_terraform_version.sh [hunk 3]
```diff
@@ -9,7 +9,11 @@ source /usr/local/bin/common.sh
 current_version=$(cat /home/easy_infra/.tfenv/version)
 output_log="/var/log/$(basename "${BASH_SOURCE[0]%%.sh}.log")"
 
-if [[ "${TERRAFORM_VERSION}" == "${current_version}" ]]; then
+if [[ -z "${TERRAFORM_VERSION}" ]]; then
```
Comment (JonZeolla) [inline build/hooks/40-set_default_terraform_version.sh:12]: This is a (very) minor bugfix.  Without it, if you set `TERRAFORM_VERSION` to an empty string, you will get a "Unable to install Terraform , continuing" message.  The reason why is because if you run `tfenv install` without an argument (i.e. `TERRAFORM_VERSION` is empty string), it exits 0.

File: docs/Technical Details/index.rst [hunk 4]
```diff
@@ -27,6 +27,10 @@ Here is a ficticious ``easy_infra.yml`` that concisely demonstrates the various
       - command: terraform validate
         description: validation
     commands:
+      fluent-bit:
```
Comment (JonZeolla) [inline docs/Technical Details/index.rst:30]: This demonstrates the new `helper: ["all"]` config option

File: docs/Technical Details/index.rst [hunk 5]
```diff
@@ -172,8 +177,22 @@ Version Argument
 ``version_argument`` is a way for us to describe how a command requests its version inside of ``easy_infra``. This is useful to know because we avoid running
 security scans (and validation, if any is specified) when the version of a tool is being queried inside of an ``easy_infra`` container.
 
+build/
+======
+
+``Dockerfile``s must all be able to be built independently, as long as their pre-requisites are met. Typically this means you pass in the appropriate
```
Comment (JonZeolla) [inline docs/Technical Details/index.rst:183]: This is additional documentation, possibly overlapping with `CONTRIBUTING.md`.  Open to de-duplicating

File: .github/workflows/commit.yml [hunk 6]
```diff
@@ -52,27 +88,27 @@ jobs:
...
-          path: vulns.${{ matrix.stage }}.*.json
+          name: Vulns - tool=${{ matrix.tool }}, environment=${{ matrix.environment }}
+          path: vulns.*.json
           if-no-files-found: error
       - name: Run tests
         run: |
           find tests -mindepth 1 -type d -exec chmod o+w {} \;
-          pipenv run invoke test --stage=${{ matrix.stage }}
+          pipenv run invoke test --tool=${{ matrix.tool }} --environment=${{ matrix.environment }}
   bump-version:
```
Comment (JonZeolla) [inline .github/workflows/commit.yml:112]: From here on it's all 🤞 because this doesn't run until it's merged.

File: build/docker-entrypoint.sh [hunk 7]
```diff
@@ -42,8 +42,8 @@ if [ "$#" -eq 0 ]; then
   if [ -x "$(which ansible)" ]; then
     echo -e "ansible\t\t ${ANSIBLE_VERSION}"
   fi
-  current_version=$(cat /home/easy_infra/.tfenv/version)
   if [ -x "$(which terraform)" ]; then
+    current_version=$(cat /home/easy_infra/.tfenv/version)
```
Comment (JonZeolla) [inline build/docker-entrypoint.sh:46]: This fixes a bug we never noticed before because `terraform` in all images, the `cat` would fail if `tfenv` wasn't installed and setup because `tfenv`/`terraform` isn't installed.

File: build/functions.j2 [hunk 8]
```diff
@@ -32,7 +32,7 @@ function process_command_exit_status() {
 {% macro function(function, file_extensions, security_tools, allow_filter, version_argument, validations) %}
 function {{ function }}() {
   arguments=("${@}")
-  easy_infra_{{ function | replace("-", "_") }}_tools=({% for security_tool in security_tools %}{{ security_tool | replace("-", "_") }}{% if not loop.last %} {% endif %}{% endfor %})
+  easy_infra_{{ function | replace("-", "_") }}_security_tools=({% for security_tool in security_tools %}{{ security_tool | replace("-", "_") }}{% if not loop.last %} {% endif %}{% endfor %})
```
Comment (JonZeolla) [inline build/functions.j2:35]: This is for clarity; we now have consistenc throughout the project on what is a `tool` vs a `security_tool` vs a `helper` and a `tag` vs `version`, etc.  More details in the original comment of this PR

File: docs/awscli/index.rst [hunk 9]
```diff
@@ -1,3 +0,0 @@
-AWS CLI
```
Comment (JonZeolla) [inline docs/awscli/index.rst:1]: Removing for now, can readd this when we add `aws` security (i.e. a cloudformation deploy shim).  On the roadmap

File: easy_infra.yml [hunk 10]
```diff
@@ -72,13 +67,20 @@ commands:
...
     allow_filter:
     - match: exec
       position: 0
     file_extensions: *id001
+    helper: ["terraform"]
     security: *id002
     validation: *id003
     version: v3.0.0
     version_argument: --version
+environments:
```
Comment (JonZeolla) [inline easy_infra.yml:82]: This new section maps environments like `aws` and `azure` to a list of commands like `aws-cli` and `azure-cli`.  It leaves things open to more commands in the future being relevant to certain environments, as well as lets us be more clear in the code when we're talking about a concept (supporting an environment like aws) vs a specific tool (like `aws-cli`) vs a command that invokes a tool (like `aws`)

File: easy_infra/config.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
+import sys
```
Comment (JonZeolla) [inline easy_infra/config.py:1]: I broke this out to remove some cyclical imports. Nothing special happening here otherwise.

File: tests/test.py [hunk 12]
```diff
@@ -24,32 +23,61 @@
 CLIENT = docker.from_env()
 
 
-def version_arguments(*, image: str, volumes: dict, working_dir: str):
-    """Test the version arguments listed in the config"""
+def version_arguments(*, tool: str, environment: str):
```
Comment (JonZeolla) [inline tests/test.py:26]: Testing functions like this one had to get much smarter so they would only test the parts that are relevant to a given tool/environment combination.  I did attempt to remove duplicate logic, but found that there were many edge cases or needs that were slightly different and made the value of making a centralized function much lower

File: tests/test.py [hunk 13]
```diff
@@ -240,8 +320,6 @@ def run_terraform(*, image: str, final: bool = False):
     invalid_volumes = {invalid_test_dir: {"bind": working_dir, "mode": "rw"}}
     checkov_test_dir = TESTS_PATH.joinpath("terraform/tool/checkov")
     checkov_volumes = {checkov_test_dir: {"bind": working_dir, "mode": "rw"}}
-    kics_config_dir = TESTS_PATH.joinpath("terraform/tool/kics")
```
Comment (JonZeolla) [inline tests/test.py:243]: This and the below removal of `kics`-related items are because `kics` is not in the `terraform` images.

**Review by JonZeolla — COMMENTED** [review]
@digeseiso and @seanctech I left a bunch of resolved comments throughout to explain / pre-empt questions you may have.

**Comment by JonZeolla** [comment]
Recent build failures are due to https://github.com/bridgecrewio/checkov/issues/4026

We can continue review while that is worked in parallel.  We will need to update `checkov` when a fix is released

File: easy_infra.yml [hunk 14]
```diff
@@ -53,12 +45,15 @@ commands:
     version: 2.2.72
```
Comment (JonZeolla) [inline easy_infra.yml:45]:
```suggestion
    version: 2.2.125
```

**Comment by JonZeolla** [comment]
We should define some of the new terminology inside the repo. Not sure if this should be in the contributing markdown or in the docs.

- Tool: An executable file in the easy_infra user's path which perform IaC actions and has an associated security tool, as described in the easy_infra.yml used when building the image.
- Security tool: An executable file in the easy_infra user's path which is configured to perform a security  scan for an associated "tool" (see above), as configured in the easy_infra.yml file used to build the image.
- Command: The name of a package that can be installed to perform a necessary functionality. It could be a tool, a security tool, or a generic helper such as fluent-bit or envconsul.
- Alias: An executable file in the easy_infra user's path which executes the installed by a command. While "aws-cli" would be a command, "aws" would be the associated alias.
- Environment: A supported destination that a "tool" may deploy into, such as a cloud provider. An environment constitutes a bundle of commands. 

Note to self: replace commands with packages?

File: tasks.py (hunk 1 above)
Comment (digeseiso) [reply tasks.py:26]: @JonZeolla I built the image successfully on my Intel-based mac

File: build/Dockerfile.azure-cli [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+                                               apt-transport-https \
+                                               lsb-release \
+                                               gnupg \
+ && rm -rf /var/cache/apt/archives/* /var/lib/apt/lists/* /tmp/* /var/tmp/* /var/cache/debconf/*-old \
+ && curl -sL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor | tee /etc/apt/trusted.gpg.d/microsoft.gpg > /dev/null \
+ && AZ_REPO=$(lsb_release -cs) \
+ # As of 2022-10-07 "The azure-cli deb package does not support ARM64 architecture"
+ && echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" | tee /etc/apt/sources.list.d/azure-cli.list
+RUN apt-get update \
+ && apt-get -y install --no-install-recommends azure-cli=${AZURE_CLI_VERSION} \
```
Comment (digeseiso) [inline build/Dockerfile.azure-cli:28]: @JonZeolla curious why this `apt-get -y install` statement is not lumped with the one in line 17

File: build/docker-entrypoint.sh [hunk 16]
```diff
@@ -5,13 +5,23 @@
...
 
 function shutdown() {
+  # Watch the fluent-bit process
+  tail --pid="$(pidof fluent-bit)" -f /dev/null &
+
+  # Sleep for twice the service flush interval from fluent-bit.conf
+  sleep .2s
+
   # Have fluent-bit dequeue and then exit
   kill -SIGTERM "$(pidof fluent-bit)"
```
Comment (digeseiso) [inline build/docker-entrypoint.sh:19]: @JonZeolla will fluent bit always be done flushing within `.2s`? If not, does fluent-bit ignore the sigterm until it  is done flushing or just shuts down gracefully regardless of what it is doing?

Comment (JonZeolla) [reply build/docker-entrypoint.sh:19]: I asked the `fluent-bit` development team on December 7th and haven't heard back, although I did get an ack that they were going to look at my question.  [Here](https://fluent-all.slack.com/archives/C0CTQGHKJ/p1670442975403489) is the slack post

Comment (JonZeolla) [reply build/docker-entrypoint.sh:19]: I did some more testing and it seems like the reason why the logs sometimes don't make it to the remote system is because it isn't pulled in by the `tail` input plugin, not because of dequeueing properly during shutdown.  However, it's still just a theory, and it seems like using the 2x the flush interval is the best way to approach this (as opposed to an arbitrary `sleep` time).  Technically it would just need any number > 1x the flush interval if I'm right. This is based off a conversation I had with a `fluent-bit` developer at KubeCon, but they weren't actually reviewing our configs/exact scenario so I took it with a grain of salt.

File: build/Dockerfile.azure-cli (hunk 15 above)
Comment (JonZeolla) [reply build/Dockerfile.azure-cli:28]: It's for caching optimization; the prior `RUN` should cache and re-run very well, but it is most likely that the `ARG` of `AZURE_CLI_VERSION` will change over time and that would invalidate the cache, so I broke it out.

File: README.md [hunk 17]
```diff
@@ -37,31 +37,31 @@ To learn more, check out [our documentation](https://easy_infra.readthedocs.io/)
...
 What `easy_infra` does in this case is:
 
 1. Perform `terraform` validation, specifically `terraform init && terraform validate`.
-1. Run `terraform` security tools\* serially, and in alphabetical order (`checkov` and then `kics`).
+1. Run `terraform` security tools\* serially, and in alphabetical order.
 1. Run the provided `terraform` command, assuming the provided configurations were confirmed as valid and did not fail any of the security policy
    validation.
 
-\* In the minimal images, only Checkov and KICS are available
+\* In the `terraform` images, only Checkov is configured by default
```
[the remark this answers is not in the document]
Comment (JonZeolla) [reply README.md:50]: Checkov is my favourite for Terraform only; KICS has a large number of rules which results in a lot of noise, it also lacks the baselining feature that `checkov` has, which is really helpful when doing roll-outs and risk based acceptance without a lot of error-prone configurations to include/exclude rules. Finally, `kics` explicitly doesn't support custom terraform modules - see [this](https://github.com/Checkmarx/kics/blob/e8de947cdaf44e5420db61c8daadfc2caf9fc2b5/docs/platforms.md#terraform-modules).

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -9,14 +9,44 @@ If you'd like to [run the test suite](#running-tests), you will also need `grype
...
+
+```bash
+pipenv run invoke build --tool=terraform --environment=none
+```
+
+To see the list of possible tools, run the following command. If you don't specify a tool, `build` will assume that you want to build "all" of the
+tools.
+
+```bash
+pipenv run python3 -c \
```
[the remark this answers is not in the document]
Comment (JonZeolla) [reply CONTRIBUTING.md:38]: I just didn't think `pipenv run python3 -c "from easy_infra import constants; print(list(constants.TOOLS))"` was that big of a deal, especially given it isn't something someone familiar with the project would need to do, it's more of a one-time getting familiar with the project or periodically checking in to see what new tools are available kind of thing.  I really don't have a good justification otherwise lol

File: build/docker-entrypoint.sh (hunk 16 above)
Comment (digeseiso) [reply build/docker-entrypoint.sh:19]: @JonZeolla sounds good, interested in fluent-bit's response

**Review by digeseiso — APPROVED** [review]
LGTM

File: docs/Technical Details/index.rst [hunk 19]
```diff
@@ -184,6 +201,82 @@ Because ``BASH_ENV`` will ensure that ``/functions`` is loaded into the shell at
...
+
+In order for a ``Dockerfile`` and a ``Dockerfrag`` to be "linked" together, they must share the same suffix. For example, ``Dockerfrag.abc`` should
+build on top of ``Dockerfile.abc``, and it is both expected that in ``Dockerfrag.abc`` it copies files using ``COPY --from=abc ...``, and that in
+``Dockerfile.abc`` the ``FROM`` statement ends with ``... as abc``.
+
+Adding to the project
+=====================
+
+Adding a tool
+^^^^^^^^^^^^^
```
Comment (JonZeolla) [inline docs/Technical Details/index.rst:255]:
```suggestion
-------------
```

File: build/docker-entrypoint.sh (hunk 16 above)
Comment (JonZeolla) [reply build/docker-entrypoint.sh:19]: Me too.  I say we ship because I can't repo with the latest config, but if they say there's a better way to do this I'll do a follow on PR

**Review by seanctech — APPROVED** [review]
LGTM
